### ISSUE TITLE
Nvidia Nemo cuts last part of transcription 

### ISSUE THREAD

**Issue by fairwell-coding** [NVIDIA/NeMo#760]
I use a Jasper15x5dr model for inference using the Nvidia Nemo framework. The results are very reliable, however, the last part of the transcription is always cut off.

Environment used:
-> python 3.8
-> cudatoolkit=10.1
-> nemo-toolkit=10.0
-> additionally installed nemo_toolkit[asr]

OS used:
-> Windows 10 Professional x64 (locally)
-> Linux serverside

What did I attempt to do (quick summary)?
-> Use Nvidia Nemo ASR Jasper 15x5dr model to perform inference on recorded wav files.

What is the problem?
-> The last section of the transcription is always missing. When the next audio file is transcribed, the cut off part is added at the beginning.

Is there another similar topic/issue being reported here on the Nvidia forums or here on github?
-> One issue regarding Nvidia nemo inference which seems to not be based on the notebook example I based my solution on (see below for details).

I am using Nvidia Nemo Jasper 15x5dr model to transcribe recorded audio files (different input formats, but all are converted to wav 16khz, 1 channel, 16bit in the preprocessing stage). For inference I based the implementation on the Online ASR Microphone Demo provided in the official Nvidia Nemo Github repository (https://github.com/NVIDIA/NeMo/blob/master/examples/asr/notebooks/2_Online_ASR_Microphone_Demo.ipynb). Instead of using PyAudio to transcribe incoming audio streams on-the-fly I receive base64 encoded preprocessed wav files but I make use of the same underlying logic outlined in the transcribe-method which is found in the FrameASR class. I use slight overlapping (too much overlapping resulted in multiples of the same phenoms in the transcription, using a very slight overlap to parse in the wav files nets excellent results for me (discouting the cut off at the end).

Before the signal (i.e. numpy array based on the incoming wav file) is infered by the loaded model in the _decode method, the buffer is shifted to the left to make room for the next frame at the very end. When debugging and checking the buffer as well as the wav files everything seems correct. The model always receives the current frame (small portion of the current wav file) and some overlapping history (the buffer contains some information of the past).

Does anyone use one of the Nvidia Nemo models for inference of recorded audio? If so, did you base your implementation on the concept provided in this example notebook? Did you change any of the buffering? I am hesitant to change this part since I assume the framework performs best like this and this has been provided intentionally like this. I played around with a lot of audio files and all show exactly the same behavior. I am welcome to any hints. The implementation I am using is part of a service and not publicly available but if deemed helpful, I can provide the relevant parts of the implementations.

Any hint/suggestion is highly welcome!

**Comment by vsl9** [comment]
https://github.com/NVIDIA/NeMo/blob/master/examples/asr/notebooks/2_Online_ASR_Microphone_Demo.ipynb covers ASR in online (streaming) scenario where the stream can be "infinite" so that's why it is processed in short segments (frames). It is expected that ASR output (transcription) has some delay. To get last part of the transcription, just make extra calls of transcribe method without any arguments: it'll create a zero frame inside and will produce the final transcription.

However, your use case looks like offline ASR: audio clip for transcription has finite duration and it is already available. So you don't need FrameASR's buffering at all. Just pass your full audio clip into `infer_signal` function to get logits and then use `_greedy_decoder` to get transcription from logits. 

Another option is to use code from ASR service demo: https://github.com/NVIDIA/NeMo/blob/9465a139575419d120044cb23477f27f3519cdb6/examples/applications/asr_service/app/routes.py#L30

**Comment by fairwell-coding** [comment]
Thank you very much for the detailed insights. This matches with my prior testing of the cut off end being attached to the next audio file being transcribed. I will keep the online inference solution in addition since this one is also needed. 

So for purely offline translation there are two working solutions:
1. Use online inference with buffering. I applied a small overlap of 10ms (i.e. 160 samples when using a sample rate of 16kHz) since this is advised to achieve best results based on the papers I found on these topics. If you want to use an overlap like I do make sure that you stay below 20ms, too much overlap will make the decoder impossible to match the same prolonged (since they are overlapped) phonemes to the same output character resulting in multiples of the same characters. For zero framing I tested a bunch of configurations. Since the total buffer for 16kHz seems to be 72000 (one frame seems to be at the start) making it 72000-16000 = 56000. This results in needing to play at least 3 zero frames to get the full transcription if the transcription goes to the very end of the input audio file. More zero frames only waste resources.
2. The approach above works great, but introduces unnecessarily the overhead of all the buffering mechanism. The logic provided in routes.py as outlined above works great as a simpler and faster solution if only offline inference is needed.
